### PR TITLE
Implement cancel a transaction request

### LINKED ISSUE #950: Cancel transaction request endpoint

### PR DESCRIPTION
**Pull request by ripzery**
Issue/Task Number: #950 

Closes #950

# Overview

This PR adds an ability to cancel a transaction request to `ewallet_api` and `admin_api`.

If the exception like below was raised, run `mix seed --settings` should fix.

```
Request: POST /api/admin/admin.login
** (exit) an exception was raised:
    ** (FunctionClauseError) no function clause matching in NaiveDateTime.add/3
        (elixir) lib/calendar/naive_datetime.ex:273: NaiveDateTime.add(~N[2019-06-10 04:28:27.111375], nil, :second)
        (ewallet_db) lib/ewallet_db/auth_token.ex:111:
```

# Changes

- Added endpoint `/me.cancel_transaction_request` to `ewallet_api`
- Added endpoint `/transaction_request.cancel` to `admin_api`
- Added `cancelled_at` field to a transaction request

# Usage

**/me.cancel_transaction_request**
```
curl http://localhost:4000/api/client/me.cancel_transaction_request \
-H "Accept: application/vnd.omisego.v1+json" \
-H "Authorization: OMGClient `echo -n "api_key:auth_token" | base64`" \
-H "Content-Type: application/json" \
-d '{
  "formatted_id": "YOUR_TRANSACTION_REQUEST_FORMATTED_ID"
}' \
-w "\n"
```

**/transaction_request.cancel**
```
curl http://localhost:4000/api/admin/transaction_request.cancel \
-H "Accept: application/vnd.omisego.v1+json" \
-H "Authorization: OMGAdmin `echo -n "user_id:auth_token" | base64`" \
-H "Content-Type: application/json" \
-d '{
  "formatted_id": "YOUR_TRANSACTION_REQUEST_FORMATTED_ID"
}' \
-w "\n" | jq
```

# Impact

Run `mix ecto.migrate`after a deployment.

# Known Issues
- [x] Fix internal server error when consuming a canceled transaction request

### CODE REVIEW

File: apps/ewallet/lib/ewallet/web/v1/error_handler.ex [hunk 1]
```diff
@@ -239,6 +239,10 @@ defmodule EWallet.Web.V1.ErrorHandler do
       code: "transaction_request:expired",
       description: "The specified transaction request has expired."
     },
+    cancelled_transaction_request: %{
+      code: "transaction_request:cancelled",
+      description: "The specified transaction request has cancelled."
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/web/v1/error_handler.ex]:
```suggestion
      description: "The specified transaction request has been cancelled."
```

File: apps/ewallet_api/lib/ewallet_api/v1/controllers/transaction_request_controller.ex [hunk 2]
```diff
@@ -32,6 +35,30 @@ defmodule EWalletAPI.V1.TransactionRequestController do
     |> respond(conn)
   end
 
+  @spec cancel_for_user(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def cancel_for_user(conn, %{"formatted_id" => formatted_id}) do
+    with {:ok, request} <- TransactionRequestFetcher.get(formatted_id) || {:error, :unauthorized},
```
Comment (unnawut) [inline apps/ewallet_api/lib/ewallet_api/v1/controllers/transaction_request_controller.ex]:
```suggestion
    with {:ok, request} <- TransactionRequestFetcher.get(formatted_id),
```

Remove unreachable condition

File: apps/admin_api/lib/admin_api/v1/controllers/transaction_request_controller.ex [hunk 3]
```diff
@@ -95,6 +95,30 @@ defmodule AdminAPI.V1.TransactionRequestController do
     |> respond(conn)
   end
 
+  @spec cancel(Plug.Conn.t(), map()) :: Plug.Conn.t()
+  def cancel(conn, %{"formatted_id" => formatted_id}) do
+    with {:ok, request} <- TransactionRequestFetcher.get(formatted_id) || {:error, :unauthorized},
```
Comment (unnawut) [inline apps/admin_api/lib/admin_api/v1/controllers/transaction_request_controller.ex]:
```suggestion
    with {:ok, request} <- TransactionRequestFetcher.get(formatted_id),
```

Remove unreachable condition

File: apps/ewallet_api/test/ewallet_api/v1/controllers/transaction_request_controller_test.exs [hunk 4]
```diff
@@ -345,4 +348,105 @@ defmodule EWalletAPI.V1.TransactionRequestControllerTest do
...
+      response =
+        client_request("/me.cancel_transaction_request", %{
+          formatted_id: transaction_request.id
+        })
+
+      # Assert the transaction request is valid
+      assert response["success"] == true
+      assert response["data"]["id"] == transaction_request.id
+      assert response["data"]["cancelled_at"] != nil
+      assert response["data"]["status"] == "cancelled"
```
Comment (unnawut) [inline apps/ewallet_api/test/ewallet_api/v1/controllers/transaction_request_controller_test.exs]: `TransactionRequest.cancelled()`

File: apps/ewallet_api/test/ewallet_api/v1/controllers/transaction_request_controller_test.exs [hunk 5]
```diff
@@ -345,4 +348,105 @@ defmodule EWalletAPI.V1.TransactionRequestControllerTest do
...
+
+      # Assert there's 1 activity log has been inserted.
+      assert [log] = get_all_activity_logs_since(before_execution)
+
+      # Assert changes
+      assert log.target_changes == %{
+               "status" => "cancelled",
```
Comment (unnawut) [inline apps/ewallet_api/test/ewallet_api/v1/controllers/transaction_request_controller_test.exs]: `TransactionRequest.cancelled()`

File: apps/ewallet_db/lib/ewallet_db/expirers/auth_expirer.ex [hunk 6]
```diff
@@ -27,10 +27,12 @@ defmodule EWalletDB.Expirers.AuthExpirer do
   @spec get_advanced_datetime(integer) :: NaiveDateTime.t() | nil
   def get_advanced_datetime(0), do: nil
 
-  def get_advanced_datetime(second) do
+  def get_advanced_datetime(second) when is_integer(second) and not is_nil(second) do
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/expirers/auth_expirer.ex]:
```suggestion
  def get_advanced_datetime(second) when is_integer(second) do
```

Remove redundant guard, `is_integer/1` should already cover `not is_nil(..)`.

**Comment by unnawut** [comment]
Noting this discussion down for reference. Consider this scenario:

1. Provider creates a transaction request
2. User consumes the transaction request with amount override -> pending consumption created
3. Provider cancelled the transaction request
4. Provider approves the consumption from #2 -> a successful transaction occurs

This is allowed since there can be cases where a provider wishes to stop accepting new payments but allow users that are mid-way through the process to continue to the end.

File: apps/ewallet_db/lib/ewallet_db/global_role.ex [hunk 7]
```diff
@@ -110,13 +110,15 @@ defmodule EWalletDB.GlobalRole do
...
+        create: :global,
+        cancel: :self
       },
       end_user_transaction_requests: %{
         all: :global,
         get: :global,
         listen: :global,
-        create: :global
+        create: :global,
+        cancel: :self
```
Comment (T-Dnzt) [inline apps/ewallet_db/lib/ewallet_db/global_role.ex]: Same.

File: apps/ewallet_db/lib/ewallet_db/global_role.ex [hunk 8]
```diff
@@ -189,13 +191,15 @@ defmodule EWalletDB.GlobalRole do
         all: :global,
         listen: :global,
         get: :global,
-        create: :none
+        create: :none,
+        cancel: :self
```
Comment (T-Dnzt) [inline apps/ewallet_db/lib/ewallet_db/global_role.ex]: Should be none, viewer can't do anything except view stuff.

File: apps/ewallet_db/lib/ewallet_db/global_role.ex [hunk 9]
```diff
@@ -189,13 +191,15 @@ defmodule EWalletDB.GlobalRole do
...
       },
       end_user_transaction_requests: %{
         all: :global,
         listen: :global,
         get: :global,
-        create: :none
+        create: :none,
+        cancel: :self
```
Comment (T-Dnzt) [inline apps/ewallet_db/lib/ewallet_db/global_role.ex]: Same, `none`.

**Comment by ripzery** [comment]
I just added some changes, please re-review it 🙇

File: apps/ewallet/lib/ewallet/validators/transaction_consumption_validator.ex [hunk 10]
```diff
@@ -120,6 +117,45 @@ defmodule EWallet.TransactionConsumptionValidator do
     :ok
   end
 
+  defp validate_transaction_request(request, operation) do
+    expiration_reason =
+      request.expiration_reason && String.to_existing_atom(request.expiration_reason)
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/validators/transaction_consumption_validator.ex]: How about turn this into `TransactionRequest.get_expiration_reason(request)`?

File: apps/ewallet/lib/ewallet/validators/transaction_consumption_validator.ex [hunk 11]
```diff
@@ -120,6 +117,45 @@ defmodule EWallet.TransactionConsumptionValidator do
...
+
+  defp do_validate_transaction_request(condition, :confirm) do
+    case condition do
+      true ->
+        :ok
+
+      :cancelled_transaction_request ->
+        :ok
+
+      expiration_reason when is_atom(expiration_reason) ->
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/validators/transaction_consumption_validator.ex]: This one doesn't guard for `not is_nil(expiration_reason)` like the function above it?

File: apps/ewallet_db/lib/ewallet_db/transaction_request.ex [hunk 12]
```diff
@@ -301,7 +296,7 @@ defmodule EWalletDB.TransactionRequest do
       set: [
         status: @expired,
         expired_at: NaiveDateTime.utc_now(),
-        expiration_reason: "expired_transaction_request"
+        expiration_reason: @expired_transaction_request
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/transaction_request.ex]: Better use `TransactionRequest.expired_transaction_request()` consistently so we know we use the same value throughout the system.

**Pull request merged**